### PR TITLE
Improving evaluation determinism.

### PR DESCRIPTION
**Pull request by Lucaweihs**
Physics in AI2-THOR is subtly non-deterministic by default due to how Unity handles the physics update step internally. This has the unfortunate side that when you run evaluation twice (even with the same random seeds) the results may be slightly different. While completely eliminating this physics non-determinism requires more work, this PR introduces some changes that dramatically improve reproducibility in my tests.

### CODE REVIEW

File: rearrange/tasks.py [hunk 1]
```diff
@@ -414,12 +375,17 @@ def drop_objects_and_teleport_back_them_to_original_locations(
...
                     # This is not a typo here.
                     self.unshuffle_env.controller.step(
-                        "OpenObject", objectId=object_id, moveMagnitude=goal_openness
+                        "OpenObject",
+                        objectId=object_id,
+                        moveMagnitude=goal_openness,
+                        **self.unshuffle_env.physics_step_kwargs,
                     )
                     action_success = self.unshuffle_env.last_event.metadata[
                         "lastActionSuccess"
```
Comment (mattdeitke) [inline rearrange/tasks.py:391]: Any reason this is not using the new `OpenObject` with `openness`, instead the old one with `moveMagnitude`?

File: rearrange/environment.py [hunk 2]
```diff
@@ -202,6 +203,14 @@ def __init__(
         self.force_cache_reset = force_cache_reset
         self.mode = mode
         self._controller_kwargs = {} if controller_kwargs is None else controller_kwargs
+        self._enhanced_physics_determinism = enhanced_physics_determinism
+
+        self.physics_step_kwargs = {}
+        if self._enhanced_physics_determinism:
+            self.physics_step_kwargs = {
+                "actionSimulationSeconds": 0.26,
+                "fixedDeltaTime": 0.02,
```
Comment (mattdeitke) [inline rearrange/environment.py:212]: Presumably these are empirically set, or is there any intuition on these constants here?

File: rearrange/environment.py [hunk 3]
```diff
@@ -181,16 +185,13 @@ def __init__(
...
-                operators = ["<=", ">=", "==", "<", ">"]
-                if any(op in v for op in operators):
-                    op = next(op for op in operators if op in v)
-                    v = f"{op}'{v.replace(op, '')}'"
-                    correct_version = eval(f"'{ai2thor.__version__}'{v}")
-                else:
-                    correct_version = ai2thor.__version__ == v
-                if not correct_version:
-                    raise ValueError(f"Please use AI2-THOR v{REQUIRED_THOR_VERSION}")
+            if ai2thor.__version__ not in ["0.0.1", None] and version.parse(
```
Comment (mattdeitke) [inline rearrange/environment.py:188]: Why allow version `0.0.1`?

File: rearrange/environment.py [hunk 4]
```diff
@@ -700,6 +723,11 @@ def drop_held_object(self) -> bool:
             action_space=self.action_space,
             action_fn=self.drop_held_object,
             thor_action="DropHandObject",
+            default_thor_kwargs={
+                "autoSimulation": False,
+                "randomMagnitude": 0.0,
```
Comment (mattdeitke) [inline rearrange/environment.py:728]: What is this?

File: rearrange/environment.py [hunk 5]
```diff
@@ -736,6 +764,11 @@ def drop_held_object_with_snap(self) -> bool:
             if held_obj is None:
                 return False
 
+            # When dropping up an object, make it breakable.
+            self.controller.step(
+                "MakeObjectBreakable", objectId=self.held_object["objectId"]
+            )
+
```
Comment (mattdeitke) [inline rearrange/environment.py:771]: It doesn't look like breakable objects are ever being set to `Unbreakable`, and, given that a breakable object starts with `breakable=True`, why is this necessary?

File: baseline_configs/two_phase/two_phase_rgb_base.py [hunk 6]
```diff
@@ -1,11 +1,12 @@
 from abc import ABC
-from typing import Optional, Sequence, Dict, Type
+from typing import Optional, Sequence, Dict, Type, Union
 
 import gym
 import gym.spaces
-from allenact.base_abstractions.sensor import SensorSuite, Sensor, DepthSensor
+import gym.spaces
```
Comment (mattdeitke) [inline baseline_configs/two_phase/two_phase_rgb_base.py:6]: duplicate import

**Review by mattdeitke — COMMENTED** [review]
Still reviewing, but a few questions, namely what are these changes fixing and which actions are non-deterministic?

To me, this PR currently feels pretty magical, in that it's now setting a few new constants (e.g., `fixedDeltaTime=0.02`, `autoSimulationSeconds=0.26`) with each action, and then this somehow solves everything 😄

**Comment by Lucaweihs** [comment]
Basically all actions in THOR are non-deterministic by default. This is because Unity runs physics updates separately from frame updates which means that an arbitrary amount of time can have passed between executing an action and receiving the output. In practice this doesn't amount to huge differences but, when evaluating thousands of episodes, can result in different metrics being output.

The "magic" constants:
```
fixedDeltaTime=0.02
autoSimulationSeconds=0.26
```
are simply a way to fix the above source of non-determinism. Together they say that, after every action, exactly 0.26 seconds of physics should be simulated with each physics step corresponding to 0.02 seconds (0.02 is a magic constant that already exists in Unity, here we just make it explicit from the python side).

File: rearrange/sensors.py [hunk 7]
```diff
@@ -1,11 +1,12 @@
 from typing import Any, Optional, Union
 
+import gym.spaces
 import gym.spaces
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply rearrange/sensors.py]: Fixed.

File: rearrange/tasks.py (hunk 1 above)
Comment (Lucaweihs) [reply rearrange/tasks.py]: No reason, fixed.

File: rearrange/environment.py (hunk 3 above)
Comment (Lucaweihs) [reply rearrange/environment.py:188]: "0.0.1" is the version that AI2-THOR reports when using a custom build. If someone is using a custom build then they can do what they want.

File: rearrange/environment.py (hunk 4 above)
Comment (Lucaweihs) [reply rearrange/environment.py:727]: By default, `"DropHandObject"` adds some random noise to the angular momentum of the dropped object. `randomMagnitude=0.0` turns this off. `autoSimulation=False` just makes it so we control the physics updates when dropping an object.

File: rearrange/environment.py (hunk 5 above)
Comment (Lucaweihs) [reply rearrange/environment.py:770]: We do make objects unbreakable in some cases to avoid some edge cases where objects placed in their correct locations would break.

File: baseline_configs/two_phase/two_phase_rgb_base.py (hunk 6 above)
Comment (Lucaweihs) [reply baseline_configs/two_phase/two_phase_rgb_base.py]: Fixed.

File: rearrange/environment.py (hunk 5 above)
Comment (mattdeitke) [reply rearrange/environment.py:770]: I see. It happens in `TeleportObject` with a new
```python
makeTeleport=True
```
parameter.
https://github.com/allenai/ai2thor-rearrangement/blob/main/rearrange/environment.py#L789

File: rearrange/environment.py (hunk 2 above)
Comment (Lucaweihs) [reply rearrange/environment.py:212]: Yes set empirically. A `fixedDeltaTime` of 0.02 is the Unity default that has high enough resolution to produce reasonable physical trajectories in most instances.
